core/src/main/java/org/eclipse/hono/util/NotificationDeviceCommandReadyConstants.java

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.util;
+
+import java.time.Instant;
+import java.util.Optional;
+
+import org.apache.qpid.proton.message.Message;
+
+/**
+ * Constants used for command readiness notifications of a device.
+ */
+public final class NotificationDeviceCommandReadyConstants extends NotificationConstants {
+    /**
+     * Prevent construction.
+     */
+    private NotificationDeviceCommandReadyConstants() {
+    }
+
+    /**
+     * Devices that are always read to receive an upstream message can use this as value for {@link MessageHelper#APP_PROPERTY_DEVICE_TTD} to indicate it in messages.
+     */
+    public static final int TTD_VALUE_ALWAYS_READY = -1;
+
+    /**
+     * Verify if a deviceNotification event (as a JSON object that is defined in the Event API) is signalling that the
+     * device for which it was received should be ready to receive a command.
+     * This is evaluated at the point in time this method is invoked.
+     *
+     * @param tenantId Tenant this message is scoped to.
+     * @param deviceNotification Message that is evaluated.
+     * @return Boolean {@code true} if the event signals that the device now should be ready to receive a command, {@code false}
```

Comment:
this method currently doesn't return a Boolean (but it should IMHO). There is also no apparent reason for passing in a tenant ID FMPOV. in fact, the method seems to be nothing more like 

```
return Optional.ofNullable(MessageHelper.getTimeUntilDisconnect(deviceNotification)).map(ttd -> {
  if (ttd == TTD_VALUE_ALWAYS_READY) {
    return Boolean.TRUE;  
  } else if (deviceNotification.getCreationTime() == 0) {
    return Boolean.FALSE;
  } else {
    final Instant creationTime = Instant.ofEpochMilli(deviceNotification.getCreationTime());
    return Instant.now().isBefore(creationTime.plusSeconds(ttd));
}).orElse(Boolean.FALSE);
```

Reply:
I will split the code to
- make the evaluation if a message indicates command readiness (returning a Boolean like you outlined above)
- and to construct a callback object that is provided to the application. 

For this second method the tenantId is needed - it cannot be retrieved from the message AFAIK. Or do I miss something?
The application itself DOES know its tenantId and provides it, so FMPOV there is no problem here (concerns?).

Reply:
> For this second method the tenantId is needed - it cannot be retrieved from the message AFAIK. Or do I miss something?

In fact we do add it to the message annotations (see `MessageHelper.annotate()`). However, this is currently done as part of the filters of HonoMessaging, i.e. when we connect the adapters to the AMQP Network directly, this would not happen. But this might only mean that we should add the annotations in the protocol adapters themselves ...

Reply:
Right - if we do that the tenantId can be eliminated. Which would be definitely better for the usability inside the application code.
I will have a look into it and if it is simple to achieve, I will eliminate tenantId from the signature here.

Reply:
I looked into it now - the `tenantId` is currently NOT annotated at the message.
If I see things correctly,  the `to` property of the AMQP message is the only part that contains the tenant, but as part of the endpoint (like `telemetry/DEFAULT_TENANT`).

Technically I think it is correct that the tenantId is not contained, since the endpoint is already scoped to it.

IMHO we have several options here:
- annotate the message with the tenantId (one additional property at each message)
- extend the `MessageConsumer` interface by a `getTenant()` method and implicitly get the tenant internally
- parse it from the endpoint (ugly :-1: )
- leave the signature like it is

For simplicitly I would currently vote for option 1 - (message contains tenantId in application property, just as deviceId).
Any concerns (my own concern would be that we continue to increase the message by more and more properties, which increases traffic and may reduce performance)  ?

Reply:
All downstream telemetry and event messages get the tenant ID and device ID added to the message annotations. We even verify that this is the case (see `org.eclipse.hono.tests.client.ClientTestBase.assertMessageProperties()`

Reply:
Yes, you are right - the tenantId is present in the annotations (not in the application properties, this was the wrong assumption).
I can eliminate the tenantId in the signature then :+1: 

BTW: why do we double the `deviceId` currently in the message annotations *and* in the application properties?

Reply:
the idea was to have all information relevant for e.g. routing the message in one place (in the annotations) and not requiring e.g. a router to gather the information from arbitrary different places.